File: E3Box/src/com/cubic/e3box/pstn/PstnService.java
package com.cubic.e3box.pstn;

import java.util.Timer;
import java.util.TimerTask;
import com.cubic.e3box.ICallCenter;

import android.os.IBinder;
import android.os.RemoteException;
import android.os.ServiceManager;
import android.util.Log;


public class PstnService{
	public static final String TAG="PstnServiceJ";

	// command to service
	private static final int PSTN_CMD_ECHO = 1;
	private static final int PSTN_CMD_DAIL = 2;
	private static final int PSTN_CMD_HANGUP =3;
	private static final int PSTN_CMD_DTMF = 4;
	private static final int PSTN_CMD_ACCEPT = 5;
	private static final int PSTN_CMD_GET_STAT = 6;
		
	// command from service
	private static final int PSTN_RPT_STAT = 100;
	private static final int PSTN_RPT_INCOMMING_CALL = 101; 

	private static void logout(String log) {
		Log.e(TAG, TAG + " " + log );
	}

	private ThirNativeServiceDef.BpThirNativeService mNativeService = null;
	private ThirNativeServiceDef.BnThirNativeService mNativeServiceCallback = new ThirNativeServiceDef.BnThirNativeService() {

		// data from pstnd
		@Override
		public int sendData(byte[] data) {
			if( mCallCenter != null )
			{
				mCallCenter.onSoundData(data);
			}
			return 0;
		}

		// report from pstnd
		@Override
		public int sendCmd(int code, String info) {
			switch(code)
			{
			case PSTN_RPT_STAT:
				// do nothing
				break;

			case PSTN_RPT_INCOMMING_CALL:
				if( mCallCenter != null )
				{
					logout("PSTN_RPT_INCOMMING_CALL info=" + info);
					mCallCenter.onIncommingCall(info);
				}
				break;
			};
			return 0;
		}

		@Override
		public int regBinder(IBinder binder) throws RemoteException {
			// empty function, do nothing
			return 0;
		}
	};

	private TimerTask mTimerTask = new TimerTask() {
		@Override
		public void run() {
			// if not connected
			if(mNativeService == null){
				logout("ServiceManager.TimerTask.run: try to connect service ..");
				// init service proxy
				IBinder binder = ServiceManager.getService(ThirNativeServiceDef.DESCRIPTOR);
				if( binder == null ) {
					logout("ServiceManager.getService failed !");
					return;
				}

				mNativeService = new ThirNativeServiceDef.BpThirNativeService(binder);
				if( mNativeService == null ) {
					logout("mNativeService create Failed!");
					return;
				}

				// regiester service
				try
				{
					int ret = mNativeService.regBinder(mNativeServiceCallback);
					logout("mNativeService.doAction.regBinder ret="+ret);
				}
				catch( Exception e )
				{
					e.printStackTrace();
				}
			} else {
				// test if service still online
				// if service offline, reconnect it
				if( sendCmd(PSTN_CMD_ECHO) < 0 ) {
					logout("ServiceManager.TimerTask.run: lost service !");
					mNativeService = null;
				}
			}
		}
	};
	private Timer mTimer = new Timer();

	private ICallCenter mCallCenter = null;

	private int sendCmd( int code, String info ) {
		int ret = 0;

		if(mNativeService == null) {
			return -100;
		}
		
		try
		{
			ret = mNativeService.sendCmd(code ,info);
		}
		catch( Exception e )
		{
			e.printStackTrace();
			ret = -99;
		}
		return ret;
	};
	private int sendCmd( int code ) {
		return sendCmd(code, "");
	};

	public boolean init(ICallCenter callCenter) {
		logout("init");
		mTimer.schedule(mTimerTask, 100, 2000);
		mCallCenter = callCenter;
		return true;
	};

	public void deinit() {
		mTimer.cancel();
	};

	public boolean dial(String number) {
		if( 0 > sendCmd(PSTN_CMD_DAIL, number) )
		{
			return false;
		}
		return true;
	};

	public boolean dtmf(String number) {
		if( 0 > sendCmd(PSTN_CMD_DTMF, number) )
		{
			return false;
		}
		return true;
	};

	public void hangup() {
		sendCmd(PSTN_CMD_HANGUP);
	};

	public void accept() {
		sendCmd(PSTN_CMD_ACCEPT);
	};

	public void writeSound( byte[] data ) {
		if( mNativeService == null ) {
			return;
		}

		try {
			mNativeService.sendData(data);
		} catch (RemoteException e) {
			e.printStackTrace();
		}
	};
}

File: pstnd/PstnService.cc
/**
 * file: PstnService.cc
 * auth: thirchina
 * brif: android c++ service for PSTN
 */

#ifndef _PSTN_SERVICE_CC_
#define _PSTN_SERVICE_CC_ 1

#include "ThirAndroidLog.h"
#include "IThirNativeService.cc"
#include "PstnCaller.cc"


class PstnService : public BnThirNativeService, public IPstnUser
{
public:
	enum
	{
		// command from user
		PSTN_CMD_ECHO = 1,
		PSTN_CMD_DAIL,
		PSTN_CMD_HANGUP,
		PSTN_CMD_DTMF,
		PSTN_CMD_ACCEPT,
		PSTN_CMD_GET_STAT,
		
		// command to user
		PSTN_RPT_STAT = 100,
		PSTN_RPT_INCOMMING_CALL,
	};

private:
	BpThirNativeService* m_user;
	PstnCaller m_caller;

	void strToC16( const char* str, char16_t* str16, int len )
	{
		const char* p = str;
		char16_t*   q = str16;
		for(int i = 0; i < len && *p; i++, p++, q++ )
		{
			*q = *p;
		}
	};

	void strToC( const char16_t* str16, char* str, int len )
	{
		const char16_t* p = str16;
		char*           q = str;
		for(int i = 0; i < len && *p; i++, p++, q++ )
		{
			*q = (char)((*p)&0xFF);
		}
	};

public:
	PstnService()
		: m_user(NULL)
		, m_caller(this)
	{
        m_caller.init();
	};

	virtual ~PstnService()
	{
		m_caller.deinit();
		if(m_user)
		{
			delete m_user;
		}
	};

	// **************************** call by user ********************************

	// command from user
	virtual int32_t sendCmd(int32_t code, const char16_t* info)
	{
		static const int INFO_LEN_MAX = 128;
		char s_info[INFO_LEN_MAX+4] = {0};

		LOGD("PstnService::sendCmd(code=%d)", code);
		switch( code )
		{
		case PSTN_CMD_ECHO:
			return 0;

		case PSTN_CMD_DAIL:
			strToC(info, s_info, INFO_LEN_MAX);
			if(!m_caller.dail(s_info))
			{
				return -1;
			}
			return 0;

		case PSTN_CMD_HANGUP:
			m_caller.hangup();
			return 0;

		case PSTN_CMD_DTMF:
            strToC(info, s_info, INFO_LEN_MAX);
			m_caller.dtmf( s_info );
            return 0;

		case PSTN_CMD_ACCEPT:
			if( !m_caller.accept() )
			{
				return -1;
			}
			return 0;

		case PSTN_CMD_GET_STAT:
			return m_caller.getStat();

		default:
			break;
		};
		return -2;
	};

	// data from user
	virtual int32_t sendData(int32_t size, const void* data)
	{
        m_caller.writeSound( data, size );
		return 0;
	};

	virtual int32_t regBinder(sp<IBinder> &binder)
	{
		LOGD("PstnService::regBinder()");
		if( m_user != NULL )
		{
			delete m_user;
            m_user = NULL;
		}
        m_user = new BpThirNativeService(binder);
		return 0;
	};

	
	// ******************** call by PSTN caller *********************
	virtual bool onIncommingCall( const char *number )
	{
		static const int NUMBER_LEN_MAX = 64;
		char16_t s_number[NUMBER_LEN_MAX+4] = {0};
		
		if(m_user)
		{
			strToC16(number, s_number, NUMBER_LEN_MAX);
			if( 0 > m_user->sendCmd(PSTN_RPT_INCOMMING_CALL, s_number) )
		    {
		        delete m_user;
                m_user = NULL;
                return false;
		    }
		}
        return true;
	};

    virtual void onSound(const void* data, int size)
    {
    	if(m_user)
		{
			if( m_user->sendData(size, data) )
			{
		        delete m_user;
                m_user = NULL;
		    }
		}
    };
};

#endif // _PSTN_SERVICE_CC_
